src/megaclient.cpp

```diff
@@ -14006,6 +14006,67 @@ Node* MegaClient::nodebyfingerprint(FileFingerprint* fingerprint)
     return mFingerprints.nodebyfingerprint(fingerprint);
 }
 
+Node* MegaClient::nodebyfingerprint(LocalNode* local_node)
+{
+    const node_vector *remote_nodes =
+      mFingerprints.nodesbyfingerprint(local_node);
+
+    if (!remote_nodes || remote_nodes->empty())
+        return nullptr;
+
+    std::string local_name = local_node->localname;
+    fsaccess->local2name(&local_name);
+    
+    // Only compare metamac if the node doesn't already exist.
+    node_vector::const_iterator remote_node =
+      std::find_if(remote_nodes->begin(),
+                   remote_nodes->end(),
+                   [&](const Node *remote_node) -> bool
+                   {
+                       return local_node->type == remote_node->type
+                              && local_name == remote_node->displayname();
```

Comment:
Why do we want to check the name? If the fingerprint matches, it looks like the same data, so that's one of the cases where we want to check the metamac. Perhaps I'm missing something, but I actually see the metamac comparison very useful for those files that are updated just a few bytes, so same size+fingerprint+name applies and we skip the update. We want, in that case, to check the metamac. Please, correct me if I'm wrong.

Reply:
I agree with you.

The original commit searched for a remote node by calling nodebyfingerprint(...) rather than nodesbyfingerprint(...). If the remote node was of the same type as the local node, it would compare them by computing the local node's metamac using the remote node's key. If they were a match, that node would be returned. If they were not a match or if the metamac could not be computed, nullptr would be returned.

I altered the code (force pushed) so that we considered all remote nodes matching a given fingerprint. If we found a remote node with the same filename and node type, we would skip the metamac comparison. This change was made to be true to the original discussion regarding this issue.

I will add a link to the discussion here shortly.

I personally feel that the original implementation was more sound. That is, look for any node with a matching fingerprint then compare metamacs.

What are your thoughts?

Reply:
:thinking: I think that skipping the filename check makes sense. Imagine I have an incremental backup system and the deltas are synced up to MEGA. It's likely that a large file is only updated at some bytes, but keeping the same name and size (and CRC). Hence, the fingerprint and filename would be the same, but not the data.
Given that we are trying to patch cases like the aforementioned, I would go for the fingerprint check (plus type, but not filename). Perhaps it's not a bad idea, however, to check other matches-by-fingerprint in case the first remote node has a different type.

Reply:
I will change the code such that:
1. We search only for a remote node with a matching fingerprint and node type.
2. If we are unable to find such a node, we return nullptr.
3. If we find such a node, we perform the metamac comparison.
4. If the metamac does not much or if it could not be computed, return nullptr.
5. Otherwise, return the matched remote node.